### DISCUSSION TITLE
Best way to address pension fund contributions (for a plugin)

### DISCUSSION THREAD

**Discussion by danielegobbetti** [kimai/kimai#4601] [General]
Hello community,
after using the feature for some time on my own instance I decided to release a small plugin for the freelance taxation in Italy ("Forfettario"). The plugin is here: https://codeberg.org/daniele/ItalianForfettarioRulesBundle

The VAT in this taxation scheme is 0, this is clear and covered by the plugin.

Freelancer are subject to the obligation of sending the invoices in electronic format, and the generation of a compatible XML file is also covered and working, with a number of additional fields on the customers' and on the user's side.

The plugin also supports a configurable percentage of the amount as pension fund contribution that can be charged to the customers. Right now I am doing this calculation in the twig template but this means of course that the amount as seen in Kimai will be the subtotal.

Before moving the logic of the pension fund contribution to the plugin's InvoiceCalculator (from my understanding the only place for it would be the ``getTax()`` method) I wanted to make sure that this is acceptable from the point of view of the project.

tl;dr:

In a (regulatory) scenario where the VAT is 0% and there are no taxes to charge to the customers, but a certain % of the invoice can be charged to the customer as pension fund contribution, is it correct to add this pension fund in ``getTax()``?

**Comment by kevinpapst** [comment]
Hi!

Thanks for bringing up that important opic, which is sleeping for too long already. Your plugin touches areas that I wanted to change for a long time now.

I think it would be to start questioning status quo and changing the invoicing API.
Throughout the world we have so many legal requirements for invoices, that we might not be able to cover all of them, but there are certain changes required in the API to be more flexible, which then would allow to cover so many more regulations.

What is needed (for example, I am sure we find more if we start digging):
- Multiple tax rates (e.g. in Quebec/Canada) required
- Some countries require a signed invoice, e.g. the XML structure you mentioned
  - Swiss: https://github.com/kimai/kimai/issues/1427
  - Zugferd / InvoiceX https://github.com/kimai/kimai/issues/1396
- QR Code, eg for SEPA information: https://github.com/kimai/kimai/issues/3743

So the API should allow 
- Using "Tax line items" instead of one the hard coded methods right now
- Add "meta files" to the invoice document itself (the XML structure or generated images)
- more flexible price calculation, the current API wasn't really built with extension points in mind

We might take some inspiration from the Stripe API: https://stripe.com/docs/api/invoices

I'd be glad to start working on that and then we can see, whether moving parts into the core is better or if country specific plugins are the way forward.

**Reply by danielegobbetti** [reply]
Hello and thanks for the prompt reply!

The Italian system is byzantine enough in itself ( :-) ) and I would be glad to (try to) summarize it from a SW requirements perspective to help you making the right choices about the API.

**Reply by danielegobbetti** [reply]
After checking it a bit, this is a brief summary, if VAT would have to be supported in Italy:
- VAT % should be tied to each activity, and then also be aggregated in the invoice by the calculator
- the way the (aggregated) VAT is treated by the calculator should depend on the customer (public administration vs private sector) AND the service
- there are many (21 if I count it correctly) reasons for TAX exemptions (!!) that should be considered, most of which apply to various material goods but to some extent also to services (e.g. sanitary sector is exempted)

For this reason, and since the "Small Business" upper limit for yearly turnover is high enough (for me) I don't think I'm willing to  address the VAT handling for Italy, and I believe it might be out of scope of the project as well, since once you hit that limit you also have lots more obligations regarding bookkeeping and the following, hence it's unlikely you can use kimai for invoicing.

I would nevertheless appreciate if there was a way to address "additional amounts to be charged to customers" (as for the pension fund) in a more programmatic way, and I believe that _some_ of the aforementioned items do fall within the scope of the project (like having different VAT % per activity and not per invoice), but that's just my 2 cents.

I hope you find this very coarse overview useful, and am happy to help further if needed! :-)

**Reply by kevinpapst** [reply]
Thanks for providing these details.

I am not going to add Activity based tax rates or any specific rules. 

What I want to provide instead are better hooks for plugin authors and a more flexible API, so we can add certain rules (like discounts or this rule you are talking about) or documents (like the XMLs for digital signage).
